test/extensions/filters/listener/http_inspector/http_inspector_test.cc

```diff
@@ -30,31 +31,297 @@ class HttpInspectorTest : public testing::Test {
+    EXPECT_EQ(accepted, Network::FilterStatus::StopIteration);
+    file_event_callback_(Event::FileReadyType::Read);
+    auto status = filter_->onData(*buffer_);
+    EXPECT_EQ(status, Network::FilterStatus::Continue);
+    if (alpn == Http::Utility::AlpnNames::get().Http11) {
+      EXPECT_EQ(1, cfg_->stats().http11_found_.value());
+    } else if (alpn == Http::Utility::AlpnNames::get().Http10) {
+      EXPECT_EQ(1, cfg_->stats().http10_found_.value());
+    } else if (alpn == Http::Utility::AlpnNames::get().Http2c) {
+      EXPECT_EQ(1, cfg_->stats().http2_found_.value());
+    } else {
+      EXPECT_EQ(alpn, "unknow alpn");
     }
   }
 
+  void testHttpInspectNotFound(absl::string_view header, bool http2 = false) {
+    init();
+    std::vector<uint8_t> data = Hex::decode(std::string(header));
+#ifdef WIN32
+    if (http2) {
+      EXPECT_CALL(os_sys_calls_, readv(_, _, _))
+          .WillOnce(
+              Invoke([&data](os_fd_t fd, const iovec* iov, int iovcnt) -> Api::SysCallSizeResult {
+                ASSERT(iov->iov_len >= data.size());
+                memcpy(iov->iov_base, data.data(), data.size());
+                return Api::SysCallSizeResult{ssize_t(data.size()), 0};
+              }))
+          .WillOnce(Return(Api::SysCallSizeResult{ssize_t(-1), SOCKET_ERROR_AGAIN}));
+    } else {
+      EXPECT_CALL(os_sys_calls_, readv(_, _, _))
+          .WillOnce(
+              Invoke([&header](os_fd_t fd, const iovec* iov, int iovcnt) -> Api::SysCallSizeResult {
+                ASSERT(iov->iov_len >= header.size());
+                memcpy(iov->iov_base, header.data(), header.size());
+                return Api::SysCallSizeResult{ssize_t(header.size()), 0};
+              }))
+          .WillOnce(Return(Api::SysCallSizeResult{ssize_t(-1), SOCKET_ERROR_AGAIN}));
+    }
+#else
+    if (http2) {
+      EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+          .WillOnce(
+              Invoke([&data](os_fd_t, void* buffer, size_t length, int) -> Api::SysCallSizeResult {
+                ASSERT(length >= data.size());
+                memcpy(buffer, data.data(), data.size());
+                return Api::SysCallSizeResult{ssize_t(data.size()), 0};
+              }));
+    } else {
+      EXPECT_CALL(os_sys_calls_, recv(42, _, _, MSG_PEEK))
+          .WillOnce(Invoke(
+              [&header](os_fd_t, void* buffer, size_t length, int) -> Api::SysCallSizeResult {
+                ASSERT(length >= header.size());
+                memcpy(buffer, header.data(), header.size());
+                return Api::SysCallSizeResult{ssize_t(header.size()), 0};
+              }));
+    }
+#endif
+    EXPECT_CALL(socket_, setRequestedApplicationProtocols(_)).Times(0);
+    auto accepted = filter_->onAccept(cb_);
+    EXPECT_EQ(accepted, Network::FilterStatus::StopIteration);
+    file_event_callback_(Event::FileReadyType::Read);
+    auto status = filter_->onData(*buffer_);
+    EXPECT_EQ(status, Network::FilterStatus::Continue);
```

Comment:
Actually, our unit tests covered the case. But due to misunderstanding the code I change the test to the wrong behavior https://github.com/envoyproxy/envoy/pull/17395/files#diff-790198a24d6c4cc17c07e52caed0fceeb7070bce698babb514412555ff78255fR323

Maybe I can valid this mistake by refactoring the test in separated PR first.